### PR TITLE
refactor!: use rustls instead of native-tls

### PR DESCRIPTION
**Pull request by kwaa**
Part of https://github.com/LemmyNet/lemmy/issues/4900

### CODE REVIEW

**Review by dullbananas — APPROVED** [review]

File: Cargo.toml
Comment (dessalines) [inline Cargo.toml:129]: I removed the default features for reqwest, and added rusttls-tls, but for some reason, `native-tls` is still in the `Cargo.lock`. 

Might need someone's help for why that's happening.

https://docs.rs/crate/reqwest/0.11.27/features

Comment (kwaa) [reply Cargo.toml:129]: May need to enable rustls feature for reqwest-middleware as well?

https://github.com/TrueLayer/reqwest-middleware/blob/4ce811911bf841db6e7d91efd65ba7e5e40a00f6/reqwest-middleware/Cargo.toml#L18

Comment (kwaa) [reply Cargo.toml:129]: Hmm... there are two versions of reqwest, of which 0.12.4 does not depend on openssl.

0.11.27 is a dependency of activitypub-federation, so this repository needs to be updated and enable rustls-tls feature.

Comment (kwaa) [reply Cargo.toml:129]: https://github.com/LemmyNet/activitypub-federation-rust/pull/118

Comment (dessalines) [reply Cargo.toml:129]: K, I'll do that for the middleware also.

Comment (Nutomic) [reply Cargo.toml:129]: Ive published version 0.5.8 of activitypub-federation which doesnt use openssl anymore.

**Review by Nutomic — APPROVED** [review]

**Comment by dessalines** [comment]
Even after this its still getting:

```
openssl v0.10.64
└── native-tls v0.2.12
    ├── hyper-tls v0.5.0
    │   └── reqwest v0.11.27
    │       ├── activitypub_federation v0.5.8
    │       │   ├── lemmy_api v0.19.5 (/home/xxx/git/lemmy/crates/api)
    │       │   │   └── lemmy_server v0.19.5 (/home/xxx/git/lemmy)
    │       │   ├── lemmy_api_common v0.19.5 (/home/xxx/git/lemmy/crates/api_common)
    │       │   │   ├── lemmy_api v0.19.5 (/home/xxx/git/lemmy/crates/api) (*)
    │       │   │   ├── lemmy_api_crud v0.19.5 (/home/xxx/git/lemmy/crates/api_crud)
    │       │   │   │   └── lemmy_server v0.19.5 (/home/xxx/git/lemmy)
    │       │   │   ├── lemmy_apub v0.19.5 (/home/xxx/git/lemmy/crates/apub)
    │       │   │   │   ├── lemmy_federate v0.19.5 (/home/xxx/git/lemmy/crates/federate)
    │       │   │   │   │   └── lemmy_server v0.19.5 (/home/xxx/git/lemmy)
    │       │   │   │   └── lemmy_server v0.19.5 (/home/xxx/git/lemmy)
    │       │   │   ├── lemmy_federate v0.19.5 (/home/xxx/git/lemmy/crates/federate) (*)
    │       │   │   ├── lemmy_routes v0.19.5 (/home/xxx/git/lemmy/crates/routes)
    │       │   │   │   └── lemmy_server v0.19.5 (/home/xxx/git/lemmy)
    │       │   │   └── lemmy_server v0.19.5 (/home/xxx/git/lemmy)
    │       │   ├── lemmy_api_crud v0.19.5 (/home/xxx/git/lemmy/crates/api_crud) (*)
    │       │   ├── lemmy_apub v0.19.5 (/home/xxx/git/lemmy/crates/apub) (*)
    │       │   ├── lemmy_db_schema v0.19.5 (/home/xxx/git/lemmy/crates/db_schema)
    │       │   │   ├── lemmy_api v0.19.5 (/home/xxx/git/lemmy/crates/api) (*)
    │       │   │   ├── lemmy_api_common v0.19.5 (/home/xxx/git/lemmy/crates/api_common) (*)
    │       │   │   ├── lemmy_api_crud v0.19.5 (/home/xxx/git/lemmy/crates/api_crud) (*)
    │       │   │   ├── lemmy_apub v0.19.5 (/home/xxx/git/lemmy/crates/apub) (*)
    │       │   │   ├── lemmy_db_views v0.19.5 (/home/xxx/git/lemmy/crates/db_views)
    │       │   │   │   ├── lemmy_api v0.19.5 (/home/xxx/git/lemmy/crates/api) (*)
    │       │   │   │   ├── lemmy_api_common v0.19.5 (/home/xxx/git/lemmy/crates/api_common) (*)
    │       │   │   │   ├── lemmy_api_crud v0.19.5 (/home/xxx/git/lemmy/crates/api_crud) (*)
    │       │   │   │   ├── lemmy_apub v0.19.5 (/home/xxx/git/lemmy/crates/apub) (*)
    │       │   │   │   └── lemmy_routes v0.19.5 (/home/xxx/git/lemmy/crates/routes) (*)
    │       │   │   ├── lemmy_db_views_actor v0.19.5 (/home/xxx/git/lemmy/crates/db_views_actor)
    │       │   │   │   ├── lemmy_api v0.19.5 (/home/xxx/git/lemmy/crates/api) (*)
    │       │   │   │   ├── lemmy_api_common v0.19.5 (/home/xxx/git/lemmy/crates/api_common) (*)
    │       │   │   │   ├── lemmy_api_crud v0.19.5 (/home/xxx/git/lemmy/crates/api_crud) (*)
    │       │   │   │   ├── lemmy_apub v0.19.5 (/home/xxx/git/lemmy/crates/apub) (*)
    │       │   │   │   ├── lemmy_federate v0.19.5 (/home/xxx/git/lemmy/crates/federate) (*)
    │       │   │   │   └── lemmy_routes v0.19.5 (/home/xxx/git/lemmy/crates/routes) (*)
    │       │   │   ├── lemmy_db_views_moderator v0.19.5 (/home/xxx/git/lemmy/crates/db_views_moderator)
    │       │   │   │   ├── lemmy_api v0.19.5 (/home/xxx/git/lemmy/crates/api) (*)
    │       │   │   │   └── lemmy_api_common v0.19.5 (/home/xxx/git/lemmy/crates/api_common) (*)
    │       │   │   ├── lemmy_federate v0.19.5 (/home/xxx/git/lemmy/crates/federate) (*)
    │       │   │   ├── lemmy_routes v0.19.5 (/home/xxx/git/lemmy/crates/routes) (*)
    │       │   │   └── lemmy_server v0.19.5 (/home/xxx/git/lemmy)
    │       │   ├── lemmy_federate v0.19.5 (/home/xxx/git/lemmy/crates/federate) (*)
    │       │   ├── lemmy_routes v0.19.5 (/home/xxx/git/lemmy/crates/routes) (*)
    │       │   └── lemmy_server v0.19.5 (/home/xxx/git/lemmy)
    │       ├── http-signature-normalization-reqwest v0.10.0
    │       │   └── activitypub_federation v0.5.8 (*)
    │       ├── lemmy_api_common v0.19.5 (/home/xxx/git/lemmy/crates/api_common) (*)
    │       ├── lemmy_apub v0.19.5 (/home/xxx/git/lemmy/crates/apub) (*)
    │       ├── lemmy_federate v0.19.5 (/home/xxx/git/lemmy/crates/federate) (*)
    │       ├── lemmy_routes v0.19.5 (/home/xxx/git/lemmy/crates/routes) (*)
    │       ├── lemmy_server v0.19.5 (/home/xxx/git/lemmy)
    │       ├── reqwest-middleware v0.2.5
    │       │   ├── activitypub_federation v0.5.8 (*)
    │       │   ├── http-signature-normalization-reqwest v0.10.0 (*)
    │       │   ├── lemmy_api_common v0.19.5 (/home/xxx/git/lemmy/crates/api_common) (*)
    │       │   ├── lemmy_routes v0.19.5 (/home/xxx/git/lemmy/crates/routes) (*)
    │       │   ├── lemmy_server v0.19.5 (/home/xxx/git/lemmy)
    │       │   ├── lemmy_utils v0.19.5 (/home/xxx/git/lemmy/crates/utils)
    │       │   │   ├── lemmy_api v0.19.5 (/home/xxx/git/lemmy/crates/api) (*)
    │       │   │   ├── lemmy_api_common v0.19.5 (/home/xxx/git/lemmy/crates/api_common) (*)
    │       │   │   ├── lemmy_api_crud v0.19.5 (/home/xxx/git/lemmy/crates/api_crud) (*)
    │       │   │   ├── lemmy_apub v0.19.5 (/home/xxx/git/lemmy/crates/apub) (*)
    │       │   │   ├── lemmy_db_schema v0.19.5 (/home/xxx/git/lemmy/crates/db_schema) (*)
    │       │   │   ├── lemmy_db_views v0.19.5 (/home/xxx/git/lemmy/crates/db_views) (*)
    │       │   │   ├── lemmy_federate v0.19.5 (/home/xxx/git/lemmy/crates/federate) (*)
    │       │   │   ├── lemmy_routes v0.19.5 (/home/xxx/git/lemmy/crates/routes) (*)
    │       │   │   └── lemmy_server v0.19.5 (/home/xxx/git/lemmy)
    │       │   └── reqwest-tracing v0.4.8
    │       │       └── lemmy_server v0.19.5 (/home/xxx/git/lemmy)
    │       ├── reqwest-tracing v0.4.8 (*)
    │       └── webmention v0.5.0
    │           └── lemmy_api_crud v0.19.5 (/home/xxx/git/lemmy/crates/api_crud) (*)
    ├── reqwest v0.11.27 (*)
    └── tokio-native-tls v0.3.1
        ├── hyper-tls v0.5.0 (*)
        └── reqwest v0.11.27 (*)
```

**Comment by kwaa** [comment]
@dessalines I think it's because it was omitted here.

https://github.com/LemmyNet/lemmy/blob/ba044c7d987acfbdedafac16cc409ba0db7709ef/Cargo.toml#L124

File: Cargo.toml
Comment (dessalines) [inline Cargo.toml:128]: @kwaa that's right here

Comment (kwaa) [reply Cargo.toml:128]: Oh, I saw it wrong.

**Comment by kwaa** [comment]
There may be other libraries using reqwest that do not enable the rustls-tls feature.

https://github.com/kwaa/lemmy/blob/854fee98e9f7481e0abebbc7405b271f8e7a4080/Cargo.lock#L4661

**Comment by kwaa** [comment]
This may be due to the `webmention` library, which does not disable the default feature.

https://github.com/timmarinin/webmention/blob/main/Cargo.toml#L17

**Comment by dessalines** [comment]
I'll do a PR there, but according to cargo tree above, there's a lot more that require it for some reason, including our own deps even (which I don't understand why).

**Comment by kwaa** [comment]
> I'll do a PR there, but according to cargo tree above, there's a lot more that require it for some reason, including our own deps even (which I don't understand why).

This should just be because they use the same version and openssl will be installed as long as one of them doesn't disable the default features.

**Comment by dessalines** [comment]
@Nutomic I tried to look up how workspace dependencies work, and tried to add `default-features = false` to all our reqwest workspace deps, but it had no affect.

Apparently workspace dep features are additive only, so none of them should require the full features of reqwest, so I don't know why all our lemmy crates are requiring openssl.

**Comment by dessalines** [comment]
That did it, thx! Okay this is ready to merge now (as long as tests pass).

I guess the messy cargo trees just had to do with circulare dependencies.

**Review by dessalines — APPROVED** [review]

**Comment by dessalines** [comment]
Oops.. I'll just do a PR to change that one dependency, and once @timmarinin merges my webmention PR, this will be handled.

**Pull request merged**
